packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -2313,18 +2383,74 @@ Hardhat Network's forking functionality only works with blocks from at least spu
 
     return { maxFeePerGas, maxPriorityFeePerGas };
   }
-}
 
-function getCommonForTracing(networkId: number, blockNumber: number): Common {
-  try {
-    const common = new Common({ chain: networkId });
+  private _selectHardfork(
+    blockNumber: BN,
+    defaultName = this._vm._common.hardfork() as HardforkName
+  ): HardforkName {
+    function assertSupported(n: HardforkName) {
+      if (!HARDHAT_NETWORK_SUPPORTED_HARDFORKS.includes(n)) {
+        throw new InternalError(`Unsupported hardfork ${n}`);
+      }
+    }
+    if (
+      this._forkBlockNumber === undefined ||
+      blockNumber.gte(new BN(this._forkBlockNumber))
+    ) {
+      assertSupported(defaultName);
+      return defaultName;
+    }
+    /** search this._hardforkActivations for the highest block number that
+     * isn't higher than blockNumber, and then return that found block number's
+     * associated hardfork name. */
+    this._assertHardforkActivations(blockNumber);
+    const activations = this._hardforkActivations!; // yes !, just asserted it.
+    let highestFound: { hardfork: HardforkName; block: number } | undefined;
+    activations.forEach((block: number, hardfork: HardforkName) => {
+      if (
+        highestFound === undefined ||
+        (block > highestFound.block && new BN(block).lte(blockNumber))
+      ) {
+        highestFound = { hardfork, block };
+      }
+    });
+    if (highestFound === undefined) {
+      throw new InternalError(
+        `Could not find a hardfork to run for block ${blockNumber}, after having looked for one in the HardhatNode's hardfork activation history, which was: ${JSON.stringify(
+          Array.from(activations.entries())
+        )}. For more information, see https://hardhat.org/hardhat-network/reference/#config`
+      );
+    }
+    assertSupported(highestFound.hardfork);
+    return highestFound.hardfork;
+  }
+
+  private _assertHardforkActivations(blockNumber: BN | number) {
+    if (this._hardforkActivations === undefined) {
+      throw new InternalError(
+        `No known hardfork for execution on historical block ${blockNumber.toString()} (relative to fork block number ${
+          this._forkBlockNumber
+        }). The node was not configured with a hardforkActivationHistory.  See http://hardhat.org/hardhat-network/guides/mainnet-forking.html#using-a-custom-hardfork-history`
+      );
+    }
+  }
 
-    common.setHardfork(common.activeHardfork(blockNumber));
+  private _getCommonForTracing(networkId: number, blockNumber: number): Common {
```

Comment:
The change here doesn't fully address what we discussed, as it still fails if you want to trace a chain that's not supported by ethereumjs

Reply:
Update: The new approach for creating commons does address this by default, as we are going to be imitating the logic of `Common.forCustomChain`. The same method would be used for standard and non-standard chains.

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1928/commits/36cbc626c470d10c4e2f6a795ad32b3ee89d9018